File: app/src/main/java/com/example/phone_contact_list/MainActivity.java
package com.example.phone_contact_list;

import android.Manifest;
import android.content.Intent;
import android.content.pm.PackageManager;
import android.database.Cursor;
import android.graphics.Bitmap;
import android.os.Build;
import android.os.Bundle;
import android.provider.ContactsContract;
import android.view.View;
import android.widget.TextView;
import android.widget.Toast;

import androidx.appcompat.app.AppCompatActivity;
import androidx.core.app.ActivityCompat;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.LinearLayoutManager;
import androidx.recyclerview.widget.RecyclerView;

import java.util.ArrayList;

public class MainActivity extends AppCompatActivity {

    private TextView tvName;
    private ArrayList<String> listDatosNombre, listDatosNumber, listDatosFoto;
    private RecyclerView recycler;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_main);

        tvName = findViewById(R.id.tvName);

        // Creamos lo necesario para el recycler view
        recycler = (RecyclerView) findViewById(R.id.rvContacts);
        recycler.setLayoutManager(new LinearLayoutManager(this, LinearLayoutManager.VERTICAL, false));
        // También podemos hacer la lista Horizontal, o usar un grid layout para que lo enseñe a modo tabla
        // (new GridLayoutManager(this, 2)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.M &&
                ContextCompat.checkSelfPermission(this, Manifest.permission.READ_CONTACTS)
                        != PackageManager.PERMISSION_GRANTED) {
            ActivityCompat.requestPermissions(this, new String[]{Manifest.permission.READ_CONTACTS}, 1);
            Toast.makeText(this, "PERMISOS QUE FUERON PEDIDOS PORQUE NO HABÍA", Toast.LENGTH_SHORT).show();
            getContacts();
        } else {
            Toast.makeText(this, "PERMISOS YA CONCENDIDOS", Toast.LENGTH_SHORT).show();
            getContacts();
        }
    }

    private void getContacts() {
        listDatosNombre = new ArrayList<String>();
        listDatosNumber = new ArrayList<String>();
        listDatosFoto = new ArrayList<String>();

        Cursor cursor = getContentResolver().query(
                ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
                null, null, null, null
        );

        if (cursor != null) {
            int testName = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME);
            int testNumber = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER);
            int foto_uri = cursor.getColumnIndex(ContactsContract.CommonDataKinds.Phone.PHOTO_URI);

            if (testName >= 0 && testNumber >= 0) {

                while (cursor.moveToNext()) {
                    String name = cursor.getString(testName);
                    String mobile = cursor.getString(testNumber);
                    String foto = cursor.getString(foto_uri);

                    listDatosNombre.add(name);
                    listDatosNumber.add(mobile);

                    if(foto!=null){
                        listDatosFoto.add(foto);
                        // Send data to the adapter
                        AdapterDatos datos = new AdapterDatos (listDatosNombre, listDatosNumber, listDatosFoto);
                        Toast.makeText(getApplicationContext(), "Rabo", Toast.LENGTH_SHORT).show();
                        recycler.setAdapter(datos);

                    } else {
                        //Load default picture
                        listDatosFoto.add("file:///android_res/drawable/vector_person.png");

                        AdapterDatos datos = new AdapterDatos (listDatosNombre, listDatosNumber, listDatosFoto);

                        datos.setOnClickListener(new View.OnClickListener() {
                            @Override
                            public void onClick(View v) {
                                int clickedPosition = recycler.getChildAdapterPosition(v);
                                // We use the child adapter to get the data where we clicked
                                openContactInfo(listDatosNombre.get(clickedPosition), listDatosNumber.get(clickedPosition));
                            }
                        });

                        recycler.setAdapter(datos);
                    }
                }
            } else {
                tvName.setText("No se encontraron las columnas necesarias en el cursor");
            }

            cursor.close();
        }
    }


    public void openContactInfo(String name, String number){
        Intent intent = new Intent(this, RecyclerViewSelectionContact.class);
        intent.putExtra("Name", name);
        intent.putExtra("Number", number);
        startActivity(intent);
    }
}

File: app/src/main/java/com/example/phone_contact_list/MainActivity.kt
package com.example.phone_contact_list

import android.Manifest
import android.content.pm.PackageManager
import android.database.Cursor
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.provider.ContactsContract
import android.provider.ContactsContract.Contacts
import android.widget.TextView
import android.widget.Toast

class MainActivity : AppCompatActivity() {

    lateinit var tvLoadContacts: TextView

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_main)

        tvLoadContacts = findViewById(R.id.tvCargarContactos)

        // ArrayList where the contacts are gonna be
        // Check build of the system, cause with the permission stuff some builds have problems
        // aka marshmallow or greater
        // We also check for the permission if it is or not granted
        if(Build.VERSION.SDK_INT >= Build.VERSION_CODES.M
            && checkSelfPermission(Manifest.permission.READ_CONTACTS)
            != PackageManager.PERMISSION_GRANTED){
            // We are gonna ask for permission cause or not granted or didnt accept
            requestPermissions(arrayOf(Manifest.permission.READ_CONTACTS),1)
            Toast.makeText(this, "PERMISOS QUE FUERON PEDIDOS PORQUE NO HABÍA", Toast.LENGTH_SHORT).show()
            getContacts()
        } else{
            Toast.makeText(this, "PERMISOS YA CONCENDIDOS", Toast.LENGTH_SHORT).show()
            // For lower marshmallow builds, we are asking the permission request
            getContacts() // Coge los permisos solo ? lol
        }
    }

    fun getContacts(): ArrayList<String> {

        var contactArrayList = ArrayList<String>()

        // To pass al the contacts to the cursor
        var cursor: Cursor? = contentResolver.query(ContactsContract.CommonDataKinds.Phone.CONTENT_URI,
        null, null,null, null)

        // Now to fetch all the contacts from cursor into the text view
        // getColumIndex = -1 si no encuentra algo

                val testName = cursor!!.getColumnIndex(ContactsContract.CommonDataKinds.Phone.DISPLAY_NAME)
                val testNumber = cursor!!.getColumnIndex(ContactsContract.CommonDataKinds.Phone.NUMBER)

                if(testName >= 0 && testNumber >=0)
                {
                    while (cursor!!.moveToNext()) {

                        val name: String = cursor.getString(testName)
                        val mobile: String = cursor.getString(testNumber)

                        val contactInfo = "$name: $mobile"
                        contactArrayList.add(contactInfo)
                    }

                    for(i in contactArrayList){
                        tvLoadContacts.text = "${tvLoadContacts.text}$i\n"
                    }

                } else {
                    tvLoadContacts.setText("No se encontraron las columnas necesarias en el cursor")
                }
            return contactArrayList
        }
    }


    /*override fun onRequestPermissionsResult(
        requestCode: Int,
        permissions: Array<out String>,
        grantResults: IntArray
    ) {
        super.onRequestPermissionsResult(requestCode, permissions, grantResults)

        if(grantResults[0] == PackageManager.PERMISSION_GRANTED)
        { // Accepted
            getContacts()
        }
    }*/
